src/machine/parsed_results.rs

```diff
@@ -1,136 +1,552 @@
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::iter::FromIterator;
+
+use super::Machine;
+use super::{HeapCellValue, Number};
 use crate::atom_table::*;
+use crate::heap_iter::{stackful_post_order_iter, NonListElider};
+use crate::machine::{F64Offset, F64Ptr, Fixnum, HeapCellValueTag};
+use crate::parser::ast::{Var, VarPtr};
+
 use dashu::*;
+use indexmap::IndexMap;
+use integer::IBig;
+use integer::UBig;
+use num_traits::cast::ToPrimitive;
 use ordered_float::OrderedFloat;
-use std::collections::BTreeMap;
-use std::collections::HashMap;
-use std::fmt::Display;
-use std::fmt::Write;
-use std::iter::FromIterator;
+use serde::ser::SerializeMap;
+use serde::ser::SerializeSeq;
+use serde::Serialize;
+use serde::Serializer;
 
 pub type QueryResult = Result<QueryResolution, String>;
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct QueryMatch {
+    pub bindings: BTreeMap<String, Value>,
+}
+
+impl Serialize for QueryMatch {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(1))?;
+        map.serialize_entry("bindings", &self.bindings)?;
```

Comment:
Does this differ noticeably from just deriving Serialize with `#[deriver(serde::Serialize)]`?

While I would expect that derive to serialize as a struct rather than a map, I would expect that a struct maps to a JSON object the same way this does.

Reply:
I didn't even consider that, because `Value` definitely needs a custom implementation. Thanks for pointing it out, I will look into it.

Reply:
We currently do not have the `derive` feature of serde enabled. I don't think it's worth it to enable it just for this, because it would include _even more macros_ in the build of Scryer Prolog even though the manual definition is fine as it is. If there are a lot of other types that we would want to serialize and that we could just `#[derive(Serialize)]` then it would probably be worth it, but I don't see that being the case even long term.

Reply:
Are you sure? `cargo tree` appears to show the `derive` feature to be enabled. Maybe a transitive dependencies already enables it?

```
cargo tree -p serde -f "{p}: {f}"
serde v1.0.204: alloc,default,derive,serde_derive,std
└── serde_derive v1.0.204 (proc-macro): default
    ├── proc-macro2 v1.0.86: default,proc-macro
    │   └── unicode-ident v1.0.12:
    ├── quote v1.0.36: default,proc-macro
    │   └── proc-macro2 v1.0.86: default,proc-macro (*)
    └── syn v2.0.72: clone-impls,default,derive,extra-traits,fold,full,parsing,printing,proc-macro,visit,visit-mut
        ├── proc-macro2 v1.0.86: default,proc-macro (*)
        ├── quote v1.0.36: default,proc-macro (*)
        └── unicode-ident v1.0.12:
```

Reply:
Oh, right, I didn't consider transitive dependencies.

Reply:
Looking at `cargo tree -i serde  -e features`

```
$> cargo tree -i serde  -e features             
serde v1.0.204
<!-- snip -->
├── serde feature "derive"
│   ├── ciborium v0.2.2 (*)
│   ├── humantime-serde v1.1.1 (*)
│   ├── iai-callgrind-runner v0.12.1 (*)
│   └── trycmd v0.15.6 (*)
├── serde feature "serde_derive"
│   └── serde feature "derive" (*)
└── serde feature "std"
    ├── ciborium feature "std" (*)
    ├── serde feature "default" (*)
    └── serde_json feature "std" (*)
```

the `derive` feature is already activated due to transitively depending on `ciborium`, `humantime-serde`, `iai-callgrind-runner`, and `trycmd`.